osquery/tables/system/windows/intel_me.cpp

```diff
@@ -40,153 +47,892 @@ DEFINE_GUID(HECI_INTERFACE_GUID,
+  std::memcpy(&fw_version.pch_ver, base_ptr, 4U);
+  base_ptr += 4U;
 
-  results.push_back(r);
+  std::memcpy(&fw_version.vendor, base_ptr, 4U);
+  base_ptr += 4U;
+
+  std::memcpy(&fw_version.last_fw_update_status, base_ptr, 4U);
+  base_ptr += 4U;
+
+  std::memcpy(&fw_version.hw_sku, base_ptr, 4U);
+  base_ptr += 4U;
+
+  std::memcpy(&fw_version.major, base_ptr, 2U);
+  base_ptr += 2U;
+
+  std::memcpy(&fw_version.minor, base_ptr, 2U);
+  base_ptr += 2U;
+
+  std::memcpy(&fw_version.build, base_ptr, 2U);
+  base_ptr += 2U;
+
+  std::memcpy(&fw_version.revision, base_ptr, 2U);
+  base_ptr += 2U;
+
+  intel_me_info.fw_version = fw_version;
+  return osquery::Status::success();
+}
+
+osquery::Status queryFirmwareVersion(IntelMEInformation& intel_me_info,
+                                     HANDLE device) {
+  using QueryFirmwareFunction =
+      osquery::Status (*)(IntelMEInformation&, HANDLE);
+
+  // clang-format off
+  const std::unordered_map<IntelMEInformation::InterfaceType, QueryFirmwareFunction> kQueryFunctions = {
+    { IntelMEInformation::InterfaceType_0, queryFirmwareVersionForInterfaceTypes0And1 },
+    { IntelMEInformation::InterfaceType_1, queryFirmwareVersionForInterfaceTypes0And1 },
+    { IntelMEInformation::InterfaceType_2, queryFirmwareVersionForInterfaceType2 }
+  };
+  // clang-format on
+
+  auto it = kQueryFunctions.find(intel_me_info.interface_type);
+  if (it == kQueryFunctions.end()) {
+    return osquery::Status::failure("Invalid or unknown interface type");
+  }
+
+  const auto& query_function = it->second;
+  return query_function(intel_me_info, device);
+}
+
+osquery::Status queryIntelMEDeviceInterface(
+    IntelMEInformation& intel_me_info, const std::string& dev_interface_path) {
+  intel_me_info = {};
+
+  DeviceHandle device;
+  auto status = openDeviceInterface(device, dev_interface_path);
+  if (!status.ok()) {
+    return status;
+  }
+
+  status = queryHECIVersion(intel_me_info, device.get());
+  if (!status.ok()) {
+    return status;
+  }
+
+  status = connectToHECIInterface(intel_me_info, device.get());
+  if (!status.ok()) {
+    return status;
+  }
+
+  status = queryFirmwareVersion(intel_me_info, device.get());
+  if (!status.ok()) {
+    return status;
+  }
+
+  return osquery::Status::success();
+}
+} // namespace
+
+namespace tables {
+void getHECIDriverVersion(QueryData& results) {
+  results = {};
+
+  std::unordered_set<std::string> dev_path_list;
+  auto status = enumerateHECIDeviceInterfacePaths(dev_path_list);
+  if (!status.ok()) {
+    LOG(WARNING) << status.getMessage();
+    return;
+  }
+
+  if (dev_path_list.empty()) {
+    LOG(INFO) << "No Intel ME device found";
+    return;
+  }
+
+  for (const auto& dev_path : dev_path_list) {
+    IntelMEInformation intel_me_info;
+    status = queryIntelMEDeviceInterface(intel_me_info, dev_path);
+    if (!status.ok()) {
+      LOG(WARNING) << status.getMessage();
+      continue;
+    }
+
+    VLOG(1) << printIntelMEVersion(intel_me_info);
```

Comment:
This seems like a table-development or DLOG. IMO we should remove this before merging.

Reply:
The reason I've added this here is to have some debug output when the table breaks due to a hardware/driver/protocol update.

The code should be a lot more reliable now, but I would like to give a chance to users to easily extract this information if they have to report an issue in this part of the code. Would you prefer a flag instead?